staging/src/k8s.io/apiserver/pkg/admission/plugin/cel/evaluator.go

```diff
@@ -0,0 +1,237 @@
+}
+
+// evaluator implements the Evaluator interface
+type evaluator struct {
+	compilationResults []CompilationResult
+	validations        []Validation
+}
+
+func NewEvaluator(compilationResults []CompilationResult, validations []Validation) Evaluator {
+	return &evaluator{
+		compilationResults,
+		validations,
+	}
+}
+
+func convertObjectToUnstructured(obj interface{}) (*unstructured.Unstructured, error) {
+	if obj == nil || reflect.ValueOf(obj).IsNil() {
+		return &unstructured.Unstructured{Object: nil}, nil
+	}
+	ret, err := runtime.DefaultUnstructuredConverter.ToUnstructured(obj)
+	if err != nil {
+		return nil, err
+	}
+	return &unstructured.Unstructured{Object: ret}, nil
+}
+
+func objectToResolveVal(r runtime.Object) (interface{}, error) {
+	if r == nil || reflect.ValueOf(r).IsNil() {
+		return nil, nil
+	}
+	v, err := convertObjectToUnstructured(r)
+	if err != nil {
+		return nil, err
+	}
+	return v.Object, nil
+}
+
+// Evaluate evaluates the compiled CEL expressions converting them into CELEvaluations
+// errors per evaluation are returned on the Evaluation object
+func (v *evaluator) Evaluate(versionedAttr *generic.VersionedAttributes, versionedParams runtime.Object) ([]Evaluation, error) {
+	// TODO: replace unstructured with ref.Val for CEL variables when native type support is available
+	evaluations := make([]Evaluation, len(v.compilationResults))
+	var err error
+
+	oldObjectVal, err := objectToResolveVal(versionedAttr.VersionedOldObject)
+	if err != nil {
+		return nil, err
+	}
+	objectVal, err := objectToResolveVal(versionedAttr.VersionedObject)
+	if err != nil {
+		return nil, err
+	}
+	paramsVal, err := objectToResolveVal(versionedParams)
+	if err != nil {
+		return nil, err
+	}
+
+	request := createAdmissionRequest(versionedAttr.Attributes)
+	requestVal, err := convertObjectToUnstructured(request)
+	if err != nil {
+		return nil, err
+	}
+	va := &validationActivation{
+		object:    objectVal,
+		oldObject: oldObjectVal,
+		params:    paramsVal,
+		request:   requestVal.Object,
+	}
+
+	for i, compilationResult := range v.compilationResults {
+		var evaluation = &evaluations[i]
+		validation := v.validations[i]
+
+		evaluation.Validation = validation
+		if compilationResult.Error != nil {
+			evaluation.Error = errors.New(fmt.Sprintf("compilation error: %v", compilationResult.Error))
+			continue
+		}
+		if compilationResult.Program == nil {
+			evaluation.Error = errors.New("unexpected internal error compiling expression")
+			continue
+		}
+		t1 := time.Now()
+		evalResult, _, err := compilationResult.Program.Eval(va)
+		elapsed := time.Since(t1)
+		evaluation.Elapsed = elapsed
+		if err != nil {
+			evaluation.Error = errors.New(fmt.Sprintf("expression '%v' resulted in error: %v", validation.Expression, err))
+		} else {
+			evaluation.EvalResult = evalResult
+		}
+	}
+
+	return evaluations, nil
+}
+
+func createAdmissionRequest(attr admission.Attributes) *admissionv1.AdmissionRequest {
```

Comment:
nit: Add a TODO to reuse https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/request/admissionreview.go#L154 instead.

Reply:
done